diff --git a/xilinx/vc707/tcl/clocks.tcl b/xilinx/vc707/tcl/clocks.tcl
index c62158d..564d267 100644
--- a/xilinx/vc707/tcl/clocks.tcl
+++ b/xilinx/vc707/tcl/clocks.tcl
@@ -31,30 +31,30 @@ set group_cl  [get_clocks -quiet {chiplink_b2c_clock              \
                                   clk_out*_vc707_sys_clock_mmcm3}]
 set group_pci [get_clocks -quiet {userclk1 txoutclk}]
 
-create_clock -add -name JTCK      -period 100   -waveform {0 50} [get_ports {jtag_TCK}]
+# create_clock -add -name JTCK      -period 100   -waveform {0 50} [get_ports {jtag_TCK}]
 if { [llength [get_ports -quiet {ulpi_clk}]] > 0 } { create_clock -add -name ULPI_CLK  -period 16.667 [get_ports {ulpi_clk}] }
 
-set group_jtag [get_clocks -quiet {JTCK}]
+# set group_jtag [get_clocks -quiet {JTCK}]
 
 if { [llength [get_ports -quiet {ulpi_clk}]] > 0 } { set group_ulpi [get_clocks -quiet {ULPI_CLK}] }
 else {set group_ulpi {} }
 
 puts "group_mem: $group_mem"
 puts "group_sys: $group_sys"
-puts "group_pci: $group_pci"
+# puts "group_pci: $group_pci"
 puts "group_cl:  $group_cl"
-puts "group_jtag: $group_jtag"
+# puts "group_jtag: $group_jtag"
 puts "group_ulpi: $group_ulpi"
 
 set groups [list]
 if { [llength $group_mem]    > 0 } { lappend groups -group $group_mem }
 if { [llength $group_sys]    > 0 } { lappend groups -group $group_sys }
-if { [llength $group_pci]    > 0 } { lappend groups -group $group_pci }
+# if { [llength $group_pci]    > 0 } { lappend groups -group $group_pci }
 if { [llength $group_cl]     > 0 } { lappend groups -group $group_cl }
-if { [llength $group_jtag]   > 0 } { lappend groups -group $group_jtag }
+# if { [llength $group_jtag]   > 0 } { lappend groups -group $group_jtag }
 if { [llength $group_ulpi]   > 0 } { lappend groups -group $group_ulpi }
 
 puts "set_clock_groups -asynchronous $groups"
 set_clock_groups -asynchronous {*}$groups
 
-set_false_path -from [get_clocks JTCK] -to [get_clocks -of_objects [get_pins lazysys/xilinxvc707mig_1/island/blackbox/u_vc707mig4gb_mig/u_iodelay_ctrl/clk_ref_mmcm_gen.mmcm_i/CLKOUT1]]
+# set_false_path -from [get_clocks JTCK] -to [get_clocks -of_objects [get_pins lazysys/xilinxvc707mig_1/island/blackbox/u_vc707mig4gb_mig/u_iodelay_ctrl/clk_ref_mmcm_gen.mmcm_i/CLKOUT1]]
