diff --git a/src/main/scala/tile/FPU.scala b/src/main/scala/tile/FPU.scala
index ad9d328f7..a629a52a2 100644
--- a/src/main/scala/tile/FPU.scala
+++ b/src/main/scala/tile/FPU.scala
@@ -796,7 +796,7 @@ class FPU(cfg: FPUParams)(implicit p: Parameters) extends FPUModule()(p) {
     regfile(load_wb_tag) := wdata
     assert(consistent(wdata))
     if (enableCommitLog)
-      printf("f%d p%d 0x%x\n", load_wb_tag, load_wb_tag + 32.U, load_wb_data)
+      printf("f%d p%d 0x%x\n", load_wb_tag, load_wb_tag + 32.U, ieee(wdata).suggestName("rtlFuzz_fregLoadData"))
     frfWriteBundle(0).wrdst := load_wb_tag
     frfWriteBundle(0).wrenf := true.B
     frfWriteBundle(0).wrdata := ieee(wdata)
@@ -930,11 +930,12 @@ class FPU(cfg: FPUParams)(implicit p: Parameters) extends FPUModule()(p) {
   val wtypeTag = Mux(divSqrt_wen, divSqrt_typeTag, wbInfo(0).typeTag)
   val wdata = box(Mux(divSqrt_wen, divSqrt_wdata, (pipes.map(_.res.data): Seq[UInt])(wbInfo(0).pipeid)), wtypeTag)
   val wexc = (pipes.map(_.res.exc): Seq[UInt])(wbInfo(0).pipeid)
-  when ((!wbInfo(0).cp && wen(0)) || divSqrt_wen) {
+val fregWrite = ((!wbInfo(0).cp && wen(0)) || divSqrt_wen).suggestName("rtlFuzz_fregWriteEnable")
+  when (fregWrite) {
     assert(consistent(wdata))
     regfile(waddr) := wdata
     if (enableCommitLog) {
-      printf("f%d p%d 0x%x\n", waddr, waddr + 32.U, ieee(wdata))
+      printf("f%d p%d 0x%x\n", waddr, waddr + 32.U, ieee(wdata).suggestName("rtlFuzz_fregWriteData"))
     }
     frfWriteBundle(1).wrdst := waddr
     frfWriteBundle(1).wrenf := true.B
